src/build.rs

```diff
@@ -251,110 +260,234 @@ impl<T> FieldsBuilder<T> {
 }
 
 impl FieldsBuilder<NamedFields> {
-    /// Add a named field with the type of the type parameter `T`
-    pub fn field_of<T>(mut self, name: &'static str, type_name: &'static str) -> Self
+    /// Add a named field constructed using the builder.
+    pub fn field<F>(mut self, builder: F) -> Self
     where
-        T: TypeInfo + ?Sized + 'static,
+        F: Fn(
+            FieldBuilder,
+        )
+            -> FieldBuilder<field_state::NameAssigned, field_state::TypeAssigned>,
     {
-        self.fields.push(Field::named_of::<T>(name, type_name));
+        let builder = builder(FieldBuilder::new());
+        self.fields.push(builder.finalize());
         self
     }
+}
```

Comment:
The main concern I have is how the api becomes a bit clunky to use with this, all the `Fn(FieldBuilder)` makes things a bit hard to read. I'm not sure if it's a good idea or not but maybe we could have convenience methods like e.g. 

```rust
    pub fn field_of<T: TypeInfo + 'static>(self, name: &'static str, type_name: &'static str) -> Self {
        self.field(|f| f.ty::<T>().name(name).type_name(type_name))
    }
```
…to cover the common cases? Maybe it's a terrible idea.

Reply:
Could add those for use in tests and handrolled impls, but the primary consumer of this API is the derive macro which would always use the builder pattern.

This PR did originally just add the docs to those helper methods, but has changed to the builder pattern after suggestions in this review. I agree it is slightly more clunky but also more flexible and doesn't require all the permutations of the helper methods, especially when we add optional indices in #80.

Reply:
Yeah, I struggled with this too. I have no better suggestion atm. :/

Definitely agree on the flexibility. We can add the convenience methods if/when we feel we need them.